### PR TITLE
Publish packet methods and refactor server.

### PR DESCRIPTION
**Pull request by moriyoshi**
I was playing with the library and wanted to do some low-level stuff above the FS layer, but it wouldn't allow me to do because packet handlers in `Server` are tightly coupled with the implementation and it is not hookable at all.

This is an attempt to publish its detail so one can extend the server based on the existing impleentation.

### CODE REVIEW

File: server_test.go [hunk 1]
```diff
@@ -176,15 +176,15 @@ func clientServerPair(t *testing.T) (*Client, *Server) {
...
 
-type sshFxpTestBadExtendedPacket struct {
+type SSHFxpTestBadExtendedPacket struct {
 	ID        uint32
 	Extension string
 	Data      string
 }
 
-func (p sshFxpTestBadExtendedPacket) id() uint32 { return p.ID }
+func (p SSHFxpTestBadExtendedPacket) Id() uint32 { return p.ID }
```
[the remark this answers is not in the document]
Comment (moriyoshi) [reply server_test.go]: Actually I'd rather name it `GetID()` or something, but looking for some comments.

File: server_stubs.go [hunk 2]
```diff
@@ -3,12 +3,12 @@
...
 
 import (
+	"fmt"
 	"os"
 	"time"
-	"fmt"
 )
 
-func runLs(dirname string, dirent os.FileInfo) string {
+func RunLs(dirname string, dirent os.FileInfo) string {
```
[the remark this answers is not in the document]
Comment (moriyoshi) [reply server_stubs.go]: Probably there should be a better name for these utility functions.

File: server_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (puellanivis) [reply server_test.go]: No, if it should be exported then it should be `ID`.

**Comment by eikenb** [comment]
I'm afraid I agree with @puellanivis that this is expands the public API way beyond what I'd want and I know the other maintainers share similar feelings as I had proposed a smaller expansion of the public API a while back and got a similar response.

I'm not sure what you were looking to do with the changes made in this pull request. Perhaps you could bring up some of what you were looking to do as issues where we can discuss them.

**Comment by moriyoshi** [comment]
Sorry for the delayed response.

> This pollutes and expands the public API far too much, all for functionality that 99.99% of people would never need.

That sounds like a bit opinionated.  What do you guys think is unecessary change exactly?  Unless a certain set of functions and structs are exposed no matter if you want them to be private or not,  rest of the world could never reuse the implementation.

It's totally fine to me that I'd fork it and use it privately, but chances are that quite a few people tend to do the same thing and I don't really like to see all those forks around.

**Comment by moriyoshi** [comment]
IMO, `Request` based implementation is a way too abstracted to implement several functions on top of VFS. In particular I couldn't implement statvfs calls with it.  At least there're some reasons behind the proposed change.

File: server_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (moriyoshi) [reply server_test.go]: It looks like the other fields such as `Path` or `Handle` have corresponding get-prefixed accessors.  So if we stick to consistency, I guess it should be named like `getID()`.

**Comment by puellanivis** [comment]
There has not been an argument made that the change is unnecessary.

However, it is the case that the raw packet API is not something most people need in order to just Open/Create/List files. So, this sort of change would be exposing an entirely different package of functionality into _this_ package.

This would end up expanding greatly the public API, by mixing two purposes into one package. And while reading the godoc, one would have to mentally separate all of the packet API from the abstracted API, and this would then complicate the use and documentation of both APIs.

If this change has enough merit that you really thinks it should be done (which no one has argued that it does not) then it should be spun into a sub-package that provides the packet API alone.

Then this package would then use that new subpackage to provide the more abstracted Open/Create/List API. This would split both APIs into different digestible packages, provide you with the functionality you are looking for, and avoid exposing the entirety of the packet API to those who just need the abstracted API.

Best of all, because this functionality was not already exported, it _can_ be spun off into its own package now.

**Comment by moriyoshi** [comment]
While I don't really get the idea what the public API looks like, I'll try splitting out the packet API into a separate package.

**Comment by eikenb** [comment]
> I couldn't implement statvfs calls with it.

@moriyoshi .. If you would like to file an issue with your statvfs use case I'd be happy to look at it and see if I can figure out if there is a way to work it into the Request based setup. I did design it specifically to be as simple as I could possibly make it, sacrificing some level of fine control, but it still might be possible to work it in somehow.

In general I've expanded the API a good bit already and am really trying to keep it from growing that much more. I do this as much to limit the support and restrictions that public APIs incur as I am to honor the original authors who wanted to keep the public API limited.

Thanks.

**Pull request closed without merging**